parlai/tasks/taskmaster/agents.py

```diff
@@ -0,0 +1,290 @@
+            self.ep_cheat_sheet = shared['ep_cheat_sheet']
+            self.num_ex = shared['num_ex']
+        else:
+            # need to set up data from scratch
+            self.ep_cheat_sheet = {}  # Stores imp. info. about each episode
+
+            # Not all episodes have relevant examples for both USER and ASSISTANT
+            # episode_map keeps track of which episode index is useful for which speaker
+            # Need to do this otherwise might end up with a situation where we cannot return anything in action
+            self.episode_map = {}
+            self.episode_map["U"] = {}
+            self.episode_map["A"] = {}
+            self.num_ex = 0
+            data_path = tm_utils._path(opt)
+            self._setup_data(data_path, opt)
+
+        self.reset()
+
+    def _setup_data(self, data_path, opt):
+        print('loading: ' + data_path)
+        with open(data_path) as data_file:
+            self.convos = json.load(data_file)
+        # Pre-processing
+        convos_update = []
+        for convo in self.convos:
+            conversation, corrupted = tm_utils.smoothen_convo(convo, opt)
+            # Filter out single greet messages and corrupted examples
+            if len(conversation) > 1 and not corrupted:
+                actual_ep_idx = len(self.ep_cheat_sheet)
+                self.ep_cheat_sheet[actual_ep_idx] = tm_utils.gen_ep_cheatsheet(
+                    conversation
+                )
+                curr_cheatsheet = self.ep_cheat_sheet[len(self.ep_cheat_sheet) - 1]
+                # calc number of examples (done here to prevent double counting if done later)
+                self.num_ex += (
+                    curr_cheatsheet[tm_utils.USER_NUM_EX]
+                    + curr_cheatsheet[tm_utils.ASSIS_NUM_EX]
+                )
+                # User example exists
+                if curr_cheatsheet[tm_utils.USER_NUM_EX] != 0:
+                    u_idx = len(self.episode_map["U"])
+                    self.episode_map["U"][u_idx] = actual_ep_idx
+                # Assistant example exists
+                if curr_cheatsheet[tm_utils.ASSIS_NUM_EX] != 0:
+                    a_idx = len(self.episode_map["A"])
+                    self.episode_map["A"][a_idx] = actual_ep_idx
+                convos_update += [conversation]
+        self.convos = convos_update
+
+    def num_examples(self):
+        return self.num_ex
+
+    def num_episodes(self):
+        # For two passes over the data: Once to teach USER and once to teach ASSISTANT
+        return len(self.episode_map["U"]) + len(self.episode_map["A"])
+
+    def get(self, episode_idx, entry_idx):
+        starts_at_odd = False
+        if episode_idx < len(self.episode_map["U"]):
+            # USER then ASSISTANT mode [First pass]
+            true_idx = self.episode_map["U"][episode_idx]
+            conversation = self.convos[true_idx]
+            convo_cheat_sheet = self.ep_cheat_sheet[true_idx]
+            first_entry, last_entry = (
```

Comment:
are these indices? if so, it might be clearer to name `first_entry_idx` etc

Reply:
I'll do that tonight